samples/settings/src/main.c

```diff
@@ -10,9 +10,38 @@ LOG_MODULE_REGISTER(golioth_hello, LOG_LEVEL_DBG);
 #include <init.h>
 #include <net/coap.h>
 #include <net/golioth/system_client.h>
+#include <net/golioth/settings.h>
 #include <samples/common/wifi.h>
 
 static struct golioth_client *client = GOLIOTH_SYSTEM_CLIENT_GET();
+static int32_t _loop_delay_s = 5;
+
+enum golioth_settings_status on_setting(
+		const char *key,
+		const struct golioth_settings_value *value)
+{
+	LOG_DBG("Received setting: key = %s, type = %d", key, value->type);
+	if (strcmp(key, "LOOP_DELAY_S") == 0) {
+		/* This setting is expected to be numeric, return an error if it's not */
+		if (value->type != GOLIOTH_SETTINGS_VALUE_TYPE_FLOAT) {
+			return GOLIOTH_SETTINGS_VALUE_FORMAT_NOT_VALID;
+		}
+
+		/* This setting must be in range [1, 100], return an error if it's not */
+		if (value->f < 1.0f || value->f > 100.0f) {
+			return GOLIOTH_SETTINGS_VALUE_OUTSIDE_RANGE;
+		}
+
+		/* Setting has passed all checks, so apply it to the loop delay */
+		_loop_delay_s = (int32_t)value->f;
+		LOG_INF("Set loop delay to %d seconds", _loop_delay_s);
```

Comment:
Would be nice to have `int` support in addition to `float`. `float` has implicit rounding problems with high (enough) numbers. The rule I try to follow is NOT to use `float` when they are not directly beneficial. They are great for signal processing for example, assuming that MCU has FPU. However, for generic code, using `float` is like asking for trouble in corner cases. Either some libc math functions are invoked, that pull a lot of flash area to the firmware footprint or there is a rounding problem (like not getting the same value that anticipating) or with FPU enabled (users want to offload hard work to HW, right?) thread stacks are much more utilized due to more registers being dumped on thread context switch. Altough it is nice to have float, I really don't think they should be promoted, especially in case like above (int counter).

Just a comment, this is a design issue, so not a blocking thing for this PR.

Reply:
Yeah, I agree we should have an `int` type.

I initially tried to add support for it. One issue is that the server sends us a double for all numbers, so there's an extra bit of code needed to check if the double is close enough to an integer to be rounded to it before handing it off to the user. I realized when trying to do this that the minimal libc in Zephyr doesn't support some functions I would need to do this in `math.h`, such as `round` or `modf`, and I didn't really have a great way to implement that myself from scratch. And I didn't want to have to depend on the more complete `newlib` just to have these functions.

But I might give this another shot. I realize now that I can multiply the double by 1000000 to move the decimal place over by 6 places, then check to see whether it's near a multiple of 1000000, and use that information to convert to int or not.

Alternatively, we can add support for `int` in `golioth` backend, then we wouldn't have to convert in the SDK.

Reply:
> Alternatively, we can add support for int in golioth backend, then we wouldn't have to convert in the SDK.

+1 on that. I don't like the idea of rounding `double` to `int` in SDK and hiding that from user.

Reply:
Opened this ticket for the backend to encode `int`s as `int`s instead of doubles: https://golioth.atlassian.net/browse/GB-346

I won't do the double-to-int conversion. I'll add a `TODO` in code for now, to add support for `int` once that ticket is done.

Reply:
Opened this ticket for the backend to encode `int`s as `int`s instead of doubles: https://golioth.atlassian.net/browse/GB-346

I won't do the double-to-int conversion. I'll add a `TODO` in code for now, to add support for `int` once that ticket is done.